nest/core/protocols.py

```diff
@@ -0,0 +1,464 @@
+        """
+        ...
+
+    @property
+    def url(self) -> str:
+        """
+        The full request URL (or path).
+        """
+        ...
+
+    @property
+    def headers(self) -> Dict[str, str]:
+        """
+        A dictionary of header names to values.
+        """
+        ...
+
+    @property
+    def cookies(self) -> Dict[str, str]:
+        """
+        A dictionary of cookie names to values.
+        """
+        ...
+
+    @property
+    def query_params(self) -> Dict[str, str]:
+        """
+        A dictionary of query parameter names to values.
+        """
+        ...
+
+    @property
+    def path_params(self) -> Dict[str, Any]:
+        """
+        A dictionary of path parameter names to values.
+        Usually extracted from the URL pattern.
+        """
+        ...
+
+    async def body(self) -> Union[bytes, str]:
+        """
+        Return the raw request body (bytes or text).
+        """
+        ...
+
+
+@runtime_checkable
+class ResponseProtocol(Protocol):
+    """
+    Abstract representation of an HTTP response.
+    """
+
+    def set_status_code(self, status_code: int) -> None:
+        """
+        Set the HTTP status code (e.g. 200, 404, etc.).
+        """
+        ...
+
+    def set_header(self, name: str, value: str) -> None:
+        """
+        Set a single header on the response.
+        """
+        ...
+
+    def set_cookie(self, name: str, value: str, **options: Any) -> None:
+        """
+        Set a cookie on the response.
+        'options' might include expires, domain, secure, httponly, etc.
+        """
+        ...
+
+    def delete_cookie(self, name: str, **options: Any) -> None:
+        """
+        Instruct the browser to delete a cookie (by setting an expired cookie).
+        """
+        ...
+
+    def set_body(self, content: Union[str, bytes]) -> None:
+        """
+        Set the final body (string or bytes).
+        (You might add overloads for JSON or streaming in a real system.)
+        """
+        ...
+
+    def set_json(self, data: Any, status_code: Optional[int] = None) -> None:
+        """
+        Encode data as JSON, set the body, and optionally set the status code.
+        """
+        ...
+
+
+###############################################################################
+# 2. ROUTING & HTTP METHODS
+###############################################################################
+
+@runtime_checkable
+class RouteDefinition(Protocol):
+    """
+    Represents a single route definition: path, HTTP methods, etc.
+    """
+    path: str
+    http_methods: List[str]
+    endpoint: Callable[..., Any]
+    name: Optional[str]
+
+
+@runtime_checkable
+class RouterProtocol(Protocol):
+    """
+    A protocol for registering routes, websocket endpoints,
+    or other specialized routes (SSE, GraphQL, etc.).
+    """
+
+    def add_route(
+        self,
+        path: str,
+        endpoint: Callable[..., Any],
```

Comment:
Should be using ParamSpec instead of ....
for more reading [url](https://sobolevn.me/2021/12/paramspec-guide)